### PR TITLE
Guns in IF2, core IF2 improvements

### LINKED ISSUE #1867: Clicking players always triggers drag doll
# Description
If you try to click on someone the drag doll will appear every time.

Drag doll:
image

There should be a delay or a deadzone for dragging to stop single clicks unintentionally being registered as drags.

Requires #1845 to be fixed first

Also: the dragging seems to be more consistent using the host, using a client has some spots on the player which don't appear to trigger it

## Category
BUG

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Get two players (1 host and 1 client)
2. Click on a player
3. Notice the drag doll pops up even if it was a single click

### LINKED ISSUE #1918: IF2 Mouse drag bugs
# Description
1. When clicking and holding mouse somewhere else and hovering yourself while still holding it you'll invoke ragdoll drag:
GIF
It should only invoke when initial click was within object borders

2. Drag start needs a "deadzone"/false start protection of some kind. I propose not starting it if drag is shorter than 1/5 of tile. It's the most noticeable on touchscreen devices: when tapping on yourself with a bruise pack it's a 9/10 chance that you will start dragging a ragdoll instead of healing yourself

### PR DESCRIPTION
**Pull request by chairbender**
### Purpose
Refactors Weapon (now called Gun because it was only handling guns) into IF2. Weapon and RequestShootMessage are removed.

Adds new interaction types to IF2 - InventoryApply, Activate, and AimApply.

Further improvements to IF2 organization. Split interaction messages up into different types and reduced their length significantly.

Also fixed an issue where suicide shot was always melee. Now you will only melee yourself if your gun has no ammo.

Fixes another issue (related to #1749) where you couldn't shoot when mousing over things.

Adjusts overall interaction logic flow.

Fixes #1918 
Fixes #1867 

As always, documentation has also been updated and improved. In particular, note the section on precedence: https://github.com/unitystation/unitystation/wiki/Interaction-Framework-V2---How-to-Cleanly-Implement-Interactions#precedence-of-interaction-components

Also refer to the new section on the overall interaction logic flow, to get an idea of how it works now (it was changed to accommodate drag and drop better while still trying to make guns work sensibly)
https://github.com/unitystation/unitystation/wiki/Interaction-Framework-V2---How-to-Cleanly-Implement-Interactions#interaction-logic-flow

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### CODE REVIEW

**Comment by chairbender** [comment]
Note, just added another commit which fixes #1918

**Comment by chairbender** [comment]
Added documentation on the new interaction logic flow:
https://github.com/unitystation/unitystation/wiki/Interaction-Framework-V2---How-to-Cleanly-Implement-Interactions#interaction-logic-flow

File: UnityProject/Assets/Scripts/Input System/HandSlot.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+
+/// <summary>
+/// enum-like class representing a hand slot. An immutable class with constant values for each slot. It's better to use this instead of
+/// strings.
+/// </summary>
+public class HandSlot
+{
+	public static readonly HandSlot RIGHT = new HandSlot("rightHand");
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Input System/HandSlot.cs:8]: Yess, I like this

**Comment by chairbender** [comment]
this is ready for review btw

**Review by fomalsd — COMMENTED** [review]
Quality code! Really well structured and documented. Yet to test this in game, will try to find some time.

**Comment by DooblyNoobly** [comment]
Nice chairbender, I just read through the documentation. Very well thought out and structured.

**Comment by fomalsd** [comment]
Are automatic guns not implemented yet? Can't shoot P90 neither on serverplayer nor on client, regardless of intent. Revolvers work though!

Also found an inventory bug (not sure if caused by this refactor): 

- Join as high ping client, 
- Quickly pick up item and drop it several times.

Client will end up thinking item is still in inventory while it's on the floor, invisible to player. It's impossible to recover client's inventory afterwards. 

Is there some kind of prediction rollback missing?
i

**Comment by chairbender** [comment]
OK, will fix those. Pickup is from previous pr. I left the existing rollback logic in place, will see why it's not working in this case.

**Comment by chairbender** [comment]
Automatics are working Btw. I think something is wrong with that p90.

**Comment by chairbender** [comment]
Nice to see you guys like it.

Bugs should be fixed now:
- P90 had wrong ammo string.
- I am only able to reproduce the inventory desync by spamming the drop button while an item is in hand of a laggy client. I think the rollback logic of the drop interaction (which I have not touched) is the issue, because it appears there is actually no more drop prediction (though there was at one point). Server is telling client to roll back something which client never predicted. I will just remove the rollback logic since drop is entirely server side now. Eventually drop can be part of IF2 and the prediction can be implemented more easily (can add it to Pickupable I think).

**Review by fomalsd — APPROVED** [review]
All fine now!

**Pull request merged**